core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -144,288 +141,181 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
       return invokeNext(ctx, command);
    }
 
-   protected final CompletableFuture<InternalCacheEntry> retrieveFromProperSource(Object key, InvocationContext ctx,
-         FlagAffectedCommand command, boolean isWrite) throws Exception {
-      return doRetrieveFromProperSource(key, null, -1, ctx, command, isWrite);
-   }
-
-   private CompletableFuture<InternalCacheEntry> doRetrieveFromProperSource(Object key, InternalCacheEntry value,
-         int lastTopologyId, InvocationContext ctx, FlagAffectedCommand command, boolean isWrite) {
-      if (value != null)
-         return CompletableFuture.completedFuture(value);
-
-      final CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-      final int currentTopologyId = cacheTopology.getTopologyId();
+   protected final CompletableFuture<Void> remoteGet(InvocationContext ctx, AbstractDataCommand command,
+                                                     boolean isWrite) throws Exception {
+      final CacheTopology cacheTopology = checkTopologyId(command);
```

Comment:
Okay, so we have a two situations when a node blocks:
* sender has lower topology than the receiver: Currently receiver throws `OTE` - we could replace `OTE` with a `OutdatedTopologyResponse` that will carry newer routing information.
* receiver has lower topology than the sender: Receiver gives the sender the entry if it owns it, if it does not it blocks on state transfer lock (in STI, probably) until it gets the command topology. If it is not a read owner in the command topology, it returns `OTR` with owners according to the current topology.

If sender receives `OTR`, it routes the command according to this topology (as it might be higher than its current topology). Not sure if the `WAIT_FOR_FIRST_RESPONSE` should return `OTR` immediately or wait until succesful response/all nodes return `OTR`.

WDYT?